### PR TITLE
Indexing metrics

### PR DESCRIPTION
**Pull request by miiu96**
- added metrics about the indexing time, number of index operations, duration of the indexing etc.

### CODE REVIEW

File: client/transport/transport.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+// transport to collect and record metrics related to the HTTP request/response cycle.
+func (m *metricsTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	startTime := time.Now()
+	size := req.ContentLength
+	resp, err := m.transport.RoundTrip(req)
+	duration := time.Since(startTime)
+
+	valueFromCtx := req.Context().Value(request.ContextKey)
+	if valueFromCtx == nil {
+		return resp, err
```
Comment (sstanculeanu) [inline client/transport/transport.go:41]: add a test with nil value on ctx for 100% coverage?

File: core/request/requests.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+		return topicWithShardID, 0
+	}
+
+	shardIDStr := split[len(split)-1]
+	shardID, err := strconv.ParseUint(shardIDStr, 10, 32)
+	if err != nil {
+		return topicWithShardID, 0
+	}
+
+	return strings.Join(split[:len(split)-1], separator), uint32(shardID)
```
Comment (sstanculeanu) [inline core/request/requests.go:55]: perhaps extract topicIndex = 0, shardIndex = 1 to some constants, then return split[topicIndex], uint32(shardID) and use shardIndex on L49

also add some tests for these 2 methods

File: process/factory/indexerFactory.go [hunk 3]
```diff
@@ -103,6 +100,29 @@ func createElasticProcessor(args ArgsIndexerFactory) (dataindexer.ElasticProcess
...
+	argsEsClient := elasticsearch.Config{
+		Addresses:     []string{args.Url},
+		Username:      args.UserName,
+		Password:      args.Password,
+		Logger:        &logging.CustomLogger{},
+		RetryOnStatus: []int{http.StatusConflict},
+		RetryBackoff:  retryBackOff,
+	}
+
+	if args.StatusMetrics == nil {
```
Comment (sstanculeanu) [inline process/factory/indexerFactory.go:113]:
```suggestion
	if check.IfNil(args.StatusMetrics) {
```

File: client/transport/transport.go (hunk 1 above)
Comment (miiu96) [reply client/transport/transport.go:41]: added

File: process/factory/indexerFactory.go (hunk 3 above)
Comment (miiu96) [reply process/factory/indexerFactory.go:113]: done

File: core/request/requests.go (hunk 2 above)
Comment (miiu96) [reply core/request/requests.go:55]: refactored and added unit tests

**Review by sstanculeanu — APPROVED** [review]

File: client/transport/transport_test.go [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+	_, _ = transportHandler.RoundTrip(req)
+
+	metricsMap := metricsHandler.GetMetrics()
+	require.Equal(t, uint64(1), metricsMap[testTopic].OperationsCount)
+	require.Equal(t, uint64(0), metricsMap[testTopic].ErrorsCount)
+	require.Equal(t, uint64(4), metricsMap[testTopic].TotalData)
+}
+
+func TestMetricsTransport_RoundTripNoValueInContextShouldNoAddMetrics(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply client/transport/transport_test.go:54]: done.

File: process/wsindexer/indexer.go [hunk 5]
```diff
@@ -15,24 +19,35 @@ var (
 	errNilDataIndexer = errors.New("nil data indexer")
 )
 
+type ArgsIndexer struct {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply process/wsindexer/indexer.go:22]: added

File: client/transport/transport.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	return &metricsTransport{
+		statusMetrics: statusMetrics,
+		transport:     http.DefaultTransport,
+	}, nil
+}
+
+// RoundTrip implements the http.RoundTripper interface and is used as a wrapper around the underlying
+// transport to collect and record metrics related to the HTTP request/response cycle.
+func (m *metricsTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	startTime := time.Now()
```
[the remark this answers is not in the document]
Comment (miiu96) [reply client/transport/transport.go:34]: no need for a nil check here because this component is like a middleware that just extracts data from the request.

File: factory/webServerFactory.go [hunk 7]
```diff
@@ -9,6 +10,10 @@ import (
 
 // CreateWebServer will create a new instance of core.WebServerHandler
 func CreateWebServer(apiConfig config.ApiRoutesConfig, statusMetricsHandler core.StatusMetricsHandler) (core.WebServerHandler, error) {
+	if check.IfNil(statusMetricsHandler) {
```
Comment (sstanculeanu) [inline factory/webServerFactory.go:13]: not really needed here, as it is only forwarded to NewMetricsFacade

**Review by sstanculeanu — DISMISSED** [review]

File: factory/webServerFactory.go (hunk 7 above)
Comment (miiu96) [reply factory/webServerFactory.go:13]: it cam be removed.

**Review by bogdan-rosianu — DISMISSED** [review]

**Review by sstanculeanu — DISMISSED** [review]

File: client/transport/transport.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	size := req.ContentLength
+	resp, err := m.transport.RoundTrip(req)
```
Comment (iulianpascalau) [inline client/transport/transport.go:36]: I do not understand this code. Why not check the error here? Why do we return the err on L41?

File: core/request/requests.go [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+// ExtendTopicWithShardID will concatenate topic with shardID
+func ExtendTopicWithShardID(topic string, shardID uint32) string {
+	return topic + separator + fmt.Sprintf("%d", shardID)
+}
+
+// SplitTopicAndShardID will extract shard id from the provided topic
+func SplitTopicAndShardID(topicWithShardID string) (string, uint32) {
+	split := strings.Split(topicWithShardID, separator)
```
Comment (iulianpascalau) [inline core/request/requests.go:44]: we do not signal any parsing error on this function. Is this ok? If yes, why?

File: scripts/observers/.env [hunk 10]
```diff
@@ -2,7 +2,7 @@ MX_CHAIN_DEPLOY_GO_URL=https://github.com/multiversx/mx-chain-deploy-go
 MX_CHAIN_PROXY_URL=https://github.com/multiversx/mx-chain-proxy-go
 
 NODE_GO_URL="https://github.com/multiversx/mx-chain-go"
-NODE_GO_BRANCH="rc/v1.6.0"
+NODE_GO_BRANCH="indexer-with-metrics"
```
Comment (iulianpascalau) [inline scripts/observers/.env:5]: add a TODO to revert here?

File: client/elasticClient.go [hunk 11]
```diff
@@ -94,8 +95,8 @@ func (ec *elasticClient) CheckAndCreateAlias(alias string, indexName string) err
 }
 
 // DoRequest will do a request to elastic server
-func (ec *elasticClient) DoRequest(req *esapi.IndexRequest) error {
-	res, err := req.Do(context.Background(), ec.client)
+func (ec *elasticClient) DoRequest(req *esapi.IndexRequest, ctx context.Context) error {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply client/elasticClient.go:98]: done

File: client/elasticClient.go [hunk 12]
```diff
@@ -154,7 +158,7 @@ func (ec *elasticClient) DoMultiGet(ids []string, index string, withSource bool,
 }
 
 // DoQueryRemove will do a query remove to elasticsearch server
-func (ec *elasticClient) DoQueryRemove(index string, body *bytes.Buffer) error {
+func (ec *elasticClient) DoQueryRemove(index string, body *bytes.Buffer, ctx context.Context) error {
 	if err := ec.doRefresh(index); err != nil {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply client/elasticClient.go:162]: done

File: client/elasticClientScroll.go [hunk 13]
```diff
@@ -14,10 +14,11 @@ import (
 )
 
 // DoCountRequest will get the number of elements that correspond with the provided query
-func (ec *elasticClient) DoCountRequest(index string, body []byte) (uint64, error) {
+func (ec *elasticClient) DoCountRequest(index string, body []byte, ctx context.Context) (uint64, error) {
```
[the remark this answers is not in the document]
Comment (miiu96) [reply client/elasticClientScroll.go:17]: done for all the functions

File: client/transport/transport.go (hunk 8 above)
Comment (miiu96) [reply client/transport/transport.go:36]: The error will be treated at an upper level. 
This function has the role to extract information about the HTTP request duration and the size of the request.

File: client/transport/transport.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (miiu96) [reply client/transport/transport.go:34]: added

File: scripts/observers/.env (hunk 10 above)
Comment (miiu96) [reply scripts/observers/.env:7]: added

**Review by bogdan-rosianu — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]
